[Docs ] Tune docs overhaul (first part) (#22112)
ray-project/ray: doc/source/tune/faq.rst

~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

Ray Tune automatically synchronizes files and checkpoints between different remote nodes as needed.
This usually happens via SSH, but this can be a :ref:`performance bottleneck <tune-bottlenecks>`,
especially when running many trials in parallel.

Instead you should use shared storage for checkpoints so that no additional synchronization across nodes
is necessary. There are two main options.

First, you can use the :ref:`SyncConfig <tune-sync-config>` to store your
logs and checkpoints on cloud storage, such as AWS S3 or Google Cloud Storage:

.. code-block:: python

    from ray import tune

    tune.run(
        tune.durable(train_fn),
        # ...,
        sync_config=tune.SyncConfig(
            upload_dir="s3://your-s3-bucket/durable-trial/"
        )
    )

Second, you can set up a shared file system like NFS. If you do this, disable automatic trial syncing:

.. code-block:: python

    from ray import tune

    tune.run(
        train_fn,
        # ...,
        local_dir="/path/to/shared/storage",
        sync_config=tune.SyncConfig(
            # Do not sync because we are on shared storage
            syncer=None
        )
    )

> Lastly, if you still want to use ssh for trial synchronization, but are not running

Reviewer: ```suggestion
Lastly, if you still want to use SSH for trial synchronization, but are not running
```

.. code-block:: python

    from ray.tune.integration.kubernetes import NamespacedKubernetesSyncer
    sync_config = tune.SyncConfig(
        syncer=NamespacedKubernetesSyncer("ray")
    )

    tune.run(train, sync_config=sync_config)

Please note that we strongly encourage you to use one of the other two options instead, as they will
result in less overhead and don't require pods to SSH into each other.

.. _tune-debugging:

How can I debug Tune experiments locally?
~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~

By default, Tune will run hyperparameter evaluations on multiple processes.
However, if you need to debug your training process, it may be easier to do everything on a single process.
You can force all Ray functions to occur on a single process with ``local_mode`` by calling the following
before ``tune.run``.

.. code-block:: python

    ray.init(local_mode=True)

Local mode with multiple configuration evaluations will interleave computation,
so it is most naturally used when running a single configuration evaluation.

Note that ``local_mode`` has some known issues, so please read :ref:`these tips <local-mode-tips>` for more info.

.. _tune-default-search-space:

> How Do I Configure Search Spaces?

Reviewer: ```suggestion
How do I configure search spaces?
```

.. code-block:: python

    parameters = {
        "qux": tune.sample_from(lambda spec: 2 + 2),
        "bar": tune.grid_search([True, False]),
        "foo": tune.grid_search([1, 2, 3]),
        "baz": "asd",  # a constant value
    }

    tune.run(trainable, config=parameters)

By default, each random variable and grid search point is sampled once.
To take multiple random samples, add ``num_samples: N`` to the experiment config.
If `grid_search` is provided as an argument, the grid will be repeated ``num_samples`` of times.

.. code-block:: python
   :emphasize-lines: 13

    # num_samples=10 repeats the 3x3 grid search 10 times, for a total of 90 trials
    tune.run(
        my_trainable,
        name="my_trainable",
        config={
            "alpha": tune.uniform(100),
            "beta": tune.sample_from(lambda spec: spec.config.alpha * np.random.normal()),
            "nn_layers": [
                tune.grid_search([16, 64, 256]),
                tune.grid_search([16, 64, 256]),
            ],
        },
        num_samples=10
    )

Note that search spaces may not be interoperable across different search algorithms.

> For example, for many search algorithms, you will not be able to use a ``grid_search`` parameter.

Reviewer: ```suggestion
For example, for many search algorithms, you will not be able to use a ``grid_search`` or ``sample_from`` parameters.
```
